### PR TITLE
fix: use correct state root after execution

### PR DESCRIPTION
**Pull request by Thegaram**

### CODE REVIEW

File: core/trace.go [hunk 1]
```diff
@@ -322,7 +322,8 @@ func (env *TraceEnv) getTxResult(state *state.StateDB, index int, block *types.B
 	// still we have no state root for per tx, only set the head and tail
 	if index == 0 {
 		txStorageTrace.RootBefore = state.GetRootHash()
-	} else if index == len(block.Transactions())-1 {
+	}
+	if index == len(block.Transactions())-1 {
```
Comment (zimpha) [inline core/trace.go:326]: This should also be merged to `develop` branch.

**Review by zimpha — APPROVED** [review]

**Review by 0xmountaintop — APPROVED** [review]

File: core/trace.go (hunk 1 above)
Comment (0xmountaintop) [reply core/trace.go:326]: ok

**Comment by colinlyguo** [comment]
CI failed with:
```
Error: cmd/geth/chaincmd.go:461:14: ineffectual assignment to err (ineffassign)
tracesJson, err := json.Marshal(traces)
```
btw, found a typo related to trace PRs: 	
```
// BlockNumber is the hash of the block in which this transaction was skipped or nil.
BlockHash *common.Hash
```

File: cmd/geth/chaincmd.go [hunk 2]
```diff
@@ -409,47 +409,53 @@ func traceTx(ctx *cli.Context) error {
...
 	header := genesisBlock.Header()
-	block := types.NewBlockWithHeader(header).WithBody([]*types.Transaction{tx}, nil)
 
 	state, err := blockchain.StateAt(genesisBlock.Root())
 	if err != nil {
 		log.Crit("Failed to initialize state", "err", err)
 	}
 
+	gasPool := new(core.GasPool).AddGas(header.GasLimit)
+	core.ApplyTransaction(config, blockchain, &common.Address{}, gasPool, state, header, tx, &header.GasUsed, *blockchain.GetVMConfig())
```
Comment (georgehao) [inline cmd/geth/chaincmd.go]:
```
func ApplyTransaction(config *params.ChainConfig, bc ChainContext, author *common.Address, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, usedGas *uint64, cfg vm.Config) (*types.Receipt, error) {
````
Don't need check the error?

Comment (Thegaram) [reply cmd/geth/chaincmd.go]: I thought we don't. But it's better to check. Thanks.

File: core/trace.go (hunk 1 above)
Comment (Thegaram) [reply core/trace.go:326]: Fixed here: https://github.com/scroll-tech/go-ethereum/pull/508

**Pull request merged**